File: course_3/module_2/examples/clustering.py
import heapq


class UnionFind:
    """Data structure to manage clusters using union-find operations."""
    def __init__(self, n):
        """
        Initialize the Union-Find structure.

        :param n: The number of elements to manage.
        """
        # Each element starts as its own parent, indicating separate clusters.
        self.parent = list(range(n))
        # Rank is used to keep the tree flat during union operations.
        self.rank = [0] * n

    def find(self, x):
        """
        Find the root of the cluster for the given element.

        :param x: The element to find the root for.
        :return: The root of the cluster.
        """
        # Path compression: make each node point directly to the root.
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        """
        Merge two clusters by connecting their roots.

        :param x: The first element.
        :param y: The second element.
        """
        # Find the roots of both clusters.
        root_x = self.find(x)
        root_y = self.find(y)

        # If roots are different, merge them.
        if root_x != root_y:
            # Attach the smaller tree under the larger tree to keep it balanced.
            if self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            elif self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y
            else:
                # If ranks are equal, choose one root and increase its rank.
                self.parent[root_y] = root_x
                self.rank[root_x] += 1


class ClusteringAlgorithm:
    """Greedy clustering algorithm using union-find."""
    def __init__(self, points, distance_matrix, num_clusters):
        """
        Initialize the clustering algorithm.

        :param points: List of points to cluster.
        :param distance_matrix: Matrix of distances between points.
        :param num_clusters: Desired number of clusters.
        """
        self.points = points
        self.distance_matrix = distance_matrix
        self.num_clusters = num_clusters
        # Union-Find structure to manage clusters.
        self.uf = UnionFind(len(points))
        # Priority queue (heap) to store edges sorted by distance.
        self.heap = self._build_heap()

    def _build_heap(self):
        """
        Build a heap to efficiently access the smallest distances.

        :return: A heap containing all pairwise distances.
        """
        heap = []
        num_points = len(self.points)
        # Iterate over all pairs of points (i, j) and store their distances in the heap.
        for i in range(num_points):
            for j in range(i + 1, num_points):
                distance = self.distance_matrix[i][j]
                heapq.heappush(heap, (distance, i, j))  # Push (distance, point1, point2).
        return heap

    def cluster(self):
        """
        Perform clustering until the desired number of clusters is reached.

        :return: A list of clusters, where each cluster is a list of points.
        """
        current_clusters = len(self.points)  # Start with each point as its own cluster.

        # Merge clusters until the desired number of clusters is reached.
        while current_clusters > self.num_clusters:
            # Extract the smallest distance from the heap.
            distance, i, j = heapq.heappop(self.heap)

            # If the points are in different clusters, merge them.
            if self.uf.find(i) != self.uf.find(j):
                self.uf.union(i, j)
                current_clusters -= 1

        # Build the final clusters from the Union-Find structure.
        clusters = {}
        for idx, point in enumerate(self.points):
            root = self.uf.find(idx)  # Find the root of the cluster.
            if root not in clusters:
                clusters[root] = []
            clusters[root].append(point)

        # Return the clusters as a list of lists.
        return list(clusters.values())

    def get_spacing(self):
        """
        Calculate the smallest distance between points in different clusters.

        :return: The spacing value (minimum inter-cluster distance).
        """
        min_distance = float('inf')
        num_points = len(self.points)

        # Iterate over all pairs of points to find the minimum distance between clusters.
        for i in range(num_points):
            for j in range(num_points):
                # Only consider points in different clusters.
                if self.uf.find(i) != self.uf.find(j):
                    min_distance = min(min_distance, self.distance_matrix[i][j])

        return min_distance


# Example usage:
if __name__ == "__main__":
    # Define the points and their pairwise distances.
    points = ["A", "B", "C", "D", "E"]
    distance_matrix = [
        [0, 2, 6, 10, 9],
        [2, 0, 4, 8, 7],
        [6, 4, 0, 3, 5],
        [10, 8, 3, 0, 6],
        [9, 7, 5, 6, 0]
    ]
    num_clusters = 2  # Desired number of clusters.

    # Perform clustering.
    clustering = ClusteringAlgorithm(points, distance_matrix, num_clusters)
    clusters = clustering.cluster()

    # Print the resulting clusters.
    print("Clustering result:")
    for i, cluster in enumerate(clusters):
        print(f"Cluster {i + 1}: {cluster}")

    # Print the spacing between clusters.
    print("\nSpacing:", clustering.get_spacing())

File: course_3/module_2/examples/clustering.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

// Edge represents a pair of points and the distance between them.
type Edge struct {
	Distance float64 // The distance between two points.
	Point1   int     // The index of the first point.
	Point2   int     // The index of the second point.
}

// PriorityQueue implements a min-heap for edges.
type PriorityQueue []Edge

// Len returns the number of elements in the priority queue.
func (pq PriorityQueue) Len() int { return len(pq) }

// Less compares two edges based on their distances.
func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].Distance < pq[j].Distance
}

// Swap exchanges two elements in the priority queue.
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

// Push adds a new edge to the priority queue.
func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(Edge))
}

// Pop removes and returns the edge with the smallest distance.
func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

// UnionFind is a structure for managing clusters.
type UnionFind struct {
	Parent []int // Array representing the parent of each element.
	Rank   []int // Array representing the rank of each tree.
}

// NewUnionFind initializes the Union-Find structure.
func NewUnionFind(size int) *UnionFind {
	parent := make([]int, size)
	rank := make([]int, size)
	for i := range parent {
		parent[i] = i // Initially, each element is its own parent.
	}
	return &UnionFind{Parent: parent, Rank: rank}
}

// Find returns the root of the cluster for a given element using path compression.
func (uf *UnionFind) Find(x int) int {
	if uf.Parent[x] != x {
		uf.Parent[x] = uf.Find(uf.Parent[x]) // Path compression to optimize future finds.
	}
	return uf.Parent[x]
}

// Union merges two clusters by rank.
func (uf *UnionFind) Union(x, y int) {
	rootX := uf.Find(x)
	rootY := uf.Find(y)
	if rootX != rootY {
		// Merge trees based on rank to maintain balance.
		if uf.Rank[rootX] > uf.Rank[rootY] {
			uf.Parent[rootY] = rootX
		} else if uf.Rank[rootX] < uf.Rank[rootY] {
			uf.Parent[rootX] = rootY
		} else {
			uf.Parent[rootY] = rootX
			uf.Rank[rootX]++
		}
	}
}

// ClusteringAlgorithm represents the greedy clustering algorithm.
type ClusteringAlgorithm struct {
	Points          []string      // List of point names.
	DistanceMatrix  [][]float64   // Matrix of distances between points.
	NumClusters     int           // Desired number of clusters.
	UnionFind       *UnionFind    // Union-Find structure to manage clusters.
	PriorityQueue   PriorityQueue // Min-heap for edges.
	CurrentClusters int           // Current number of clusters.
	MinSpacing      float64       // Minimum spacing between clusters.
}

// validateDistanceMatrix validates the input distance matrix for correctness.
func validateDistanceMatrix(matrix [][]float64) error {
	n := len(matrix)
	for i := 0; i < n; i++ {
		if len(matrix[i]) != n {
			return fmt.Errorf("distance matrix is not square")
		}
		if matrix[i][i] != 0 {
			return fmt.Errorf("distance matrix diagonal must be zero")
		}
		for j := i + 1; j < n; j++ {
			if matrix[i][j] != matrix[j][i] {
				return fmt.Errorf("distance matrix is not symmetric")
			}
		}
	}
	return nil
}

// NewClusteringAlgorithm initializes the clustering algorithm.
func NewClusteringAlgorithm(points []string, distanceMatrix [][]float64, numClusters int) *ClusteringAlgorithm {
	if err := validateDistanceMatrix(distanceMatrix); err != nil {
		panic(err)
	}

	pq := PriorityQueue{}
	heap.Init(&pq)
	// Add all edges to the priority queue.
	for i := 0; i < len(points); i++ {
		for j := i + 1; j < len(points); j++ {
			heap.Push(&pq, Edge{Distance: distanceMatrix[i][j], Point1: i, Point2: j})
		}
	}
	return &ClusteringAlgorithm{
		Points:          points,
		DistanceMatrix:  distanceMatrix,
		NumClusters:     numClusters,
		UnionFind:       NewUnionFind(len(points)),
		PriorityQueue:   pq,
		CurrentClusters: len(points),
		MinSpacing:      math.MaxFloat64,
	}
}

// Cluster performs the clustering until the desired number of clusters is reached.
func (ca *ClusteringAlgorithm) Cluster() [][]string {
	for ca.CurrentClusters > ca.NumClusters {
		// Get the edge with the smallest distance.
		edge := heap.Pop(&ca.PriorityQueue).(Edge)
		// Merge clusters if they are not already connected.
		if ca.UnionFind.Find(edge.Point1) != ca.UnionFind.Find(edge.Point2) {
			ca.UnionFind.Union(edge.Point1, edge.Point2)
			ca.CurrentClusters--
		}
	}

	// Calculate the minimum spacing between clusters.
	ca.MinSpacing = math.MaxFloat64
	for _, edge := range ca.PriorityQueue {
		if ca.UnionFind.Find(edge.Point1) != ca.UnionFind.Find(edge.Point2) {
			if edge.Distance < ca.MinSpacing {
				ca.MinSpacing = edge.Distance
			}
		}
	}

	// Build the final clusters.
	clusters := make(map[int][]string)
	for i, point := range ca.Points {
		root := ca.UnionFind.Find(i)
		clusters[root] = append(clusters[root], point)
	}

	// Convert the map of clusters into a slice of slices.
	result := make([][]string, 0, len(clusters))
	for _, cluster := range clusters {
		result = append(result, cluster)
	}
	return result
}

// Main function to execute the clustering algorithm.
func main() {
	// Define the points and the distance matrix.
	points := []string{"A", "B", "C", "D", "E"}
	distanceMatrix := [][]float64{
		{0, 2, 6, 10, 9},
		{2, 0, 4, 8, 7},
		{6, 4, 0, 3, 5},
		{10, 8, 3, 0, 6},
		{9, 7, 5, 6, 0},
	}
	numClusters := 2 // Desired number of clusters.

	// Initialize and run the clustering algorithm.
	algorithm := NewClusteringAlgorithm(points, distanceMatrix, numClusters)
	clusters := algorithm.Cluster()

	// Print the clustering result.
	fmt.Println("Clustering result:")
	for i, cluster := range clusters {
		fmt.Printf("Cluster %d: %v\n", i+1, cluster)
	}

	// Print the minimum spacing between clusters.
	fmt.Printf("\nSpacing: %.2f\n", algorithm.MinSpacing)
}
